flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java

```diff
@@ -61,158 +61,180 @@
 	private final String clusterId;
 	private final String nameSpace;
 
-	public Fabric8FlinkKubeClient(Configuration flinkConfig, KubernetesClient client) {
+	private final ExecutorWrapper executorWrapper;
+
+	public Fabric8FlinkKubeClient(Configuration flinkConfig, KubernetesClient client, ExecutorWrapper executorWrapper) {
 		this.flinkConfig = checkNotNull(flinkConfig);
 		this.internalClient = checkNotNull(client);
 		this.clusterId = checkNotNull(flinkConfig.getString(KubernetesConfigOptions.CLUSTER_ID));
 
 		this.nameSpace = flinkConfig.getString(KubernetesConfigOptions.NAMESPACE);
+
+		this.executorWrapper = executorWrapper;
 	}
 
 	@Override
-	public void createJobManagerComponent(KubernetesJobManagerSpecification kubernetesJMSpec) {
+	public CompletableFuture<Void> createJobManagerComponent(KubernetesJobManagerSpecification kubernetesJMSpec) {
 		final Deployment deployment = kubernetesJMSpec.getDeployment();
 		final List<HasMetadata> accompanyingResources = kubernetesJMSpec.getAccompanyingResources();
 
 		// create Deployment
 		LOG.debug("Start to create deployment with spec {}", deployment.getSpec().toString());
-		final Deployment createdDeployment = this.internalClient
-			.apps()
-			.deployments()
-			.inNamespace(this.nameSpace)
-			.create(deployment);
-
-		// Note that we should use the uid of the created Deployment for the OwnerReference.
-		setOwnerReference(createdDeployment, accompanyingResources);
 
-		this.internalClient
-			.resourceList(accompanyingResources)
-			.inNamespace(this.nameSpace)
-			.createOrReplace();
+		return CompletableFuture.runAsync(() -> {
+			final Deployment createdDeployment = this.internalClient
+				.apps()
+				.deployments()
+				.inNamespace(this.nameSpace)
+				.create(deployment);
+
+			// Note that we should use the uid of the created Deployment for the OwnerReference.
+			setOwnerReference(createdDeployment, accompanyingResources);
+
+			this.internalClient
+				.resourceList(accompanyingResources)
+				.inNamespace(this.nameSpace)
+				.createOrReplace();
+		}, executorWrapper.getExecutor());
 	}
 
 	@Override
 	public void createTaskManagerPod(KubernetesPod kubernetesPod) {
-		final Deployment masterDeployment = this.internalClient
-			.apps()
-			.deployments()
-			.inNamespace(this.nameSpace)
-			.withName(KubernetesUtils.getDeploymentName(clusterId))
-			.get();
-
-		if (masterDeployment == null) {
-			throw new RuntimeException(
-				"Failed to find Deployment named " + clusterId + " in namespace " + this.nameSpace);
-		}
+		CompletableFuture.runAsync(() -> {
```

Comment:
Good point. I will add a `retry` logic here. When all the retry times(3) is exhausted, we will throw the exception here.

Reply:
I think one of `stopPod` usages is not fixed in `KubernetesResourceManager#stopWorker`. The code to stop pod by its `ResourceID` can probably be deduplicated between KubernetesResourceManager#stopWorker` and KubernetesResourceManager#removePodIfTerminated`

Reply:
Side note: I do not quite understand the purpose of `FlinkKubeClient#handleException`.
Could we not just log it in `KubernetesClusterDescriptor` and `KubernetesResourceManager`?

Reply:
For `FlinkKubeClient#handleException`, i will remove it from `KubernetesResourceManager `. Since when we make `createTaskManagerPod` and `stopPod` asynchronous, it will not throw any exception then.

Actually, the `FlinkKubeClient#handleException` do not have meaningful operations. And it will make the exception stack confusing. I think it could be removed.

Reply:
I agree we can remove `FlinkKubeClient#handleException` as a follow-up

Reply:
One more question here if `FlinkKubeClient#createTaskManagerPod` fails due to the cluster long-time unavailability. We risk to enter into a frequent pod creation retry loop without any back-off. Is it handled somehow by fabric8 or okHttpClient client? Otherwise, we might want to follow-up on this.

Reply:
Also, it would be nice to add a test that KubernetesResourceManager retries creation of the pod if `FlinkKubeClient#createTaskManagerPod` future fails, using some testing implementation of `FlinkKubeClient`.

Reply:
This is a valid concern. I am afraid `fabric8` and `okHttpClient` could not cover this situation. Regarding to a follow-up, i tend to fix it in this PR.

I will move the retry logic into the `FlinkKubeClient#createTaskManagerPod`. In `KubernetesResourceManager kubeClient.createTaskManagerPod#whenComplete`, we just log the error. This could avoid the infinite retry. So in this situation, the slot allocation will timeout in `JobMaster#SlotPool`, and it will allocate again.

WDYT?

Reply:
I checked the JM implementation and it looks that JM decides when to timeout the pending request. Judging the existing active RM integrations (e.g. yarn), the RM should retry slot allocation until JM calls `RM#cancelSlotRequest` which happens in case of JM slot request timeout. Therefore, I still think that following `KubernetesResourceManager#onError` is more correct if internal fabric8 http request fails (still needs a unit test if not added yet).

My concern is more about a back-off strategy that the `KubernetesResourceManager` does not retry `FlinkKubeClient#createTaskManagerPod` too often but there is a certain time interval between retries. Maybe, something in fabric8 `Config` (like `KUBERNETES_REQUEST_TIMEOUT_SYSTEM_PROPERTY`) already can handle this?

Reply:
Thanks a lot for your comments. I will dig into fabric8 kubernetes client for more information. Also i will drop the current retry logics in `FlinkKubeClient#createTaskManagerPod` and still keep the retry in `KubernetesResourceManager`. A unit test should be added for such situation.

Reply:
After check the `fabric8`, `okhttp` client and some information about K8s apiserver, i believe that the internal `requestTimeout` and `connectionTimeout` could completely solve the "retry too often" problems. Since the connection may be normal and the apiserver return with some error code(`403, 404, 500`, etc.) immediately. 

So i think we still need to add a retry internal in `KubernetesResourceManager`. I have leave the default value to 3s. Also i add a test `testCreateTaskManagerPodFailedAndRetry` for creating taskmanager pod failed and retry.

Reply:
Alright, I think this should work in most cases. As a follow-up, we might want to make it configurable or add back-off (similar to 'NETWORK_REQUEST_BACKOFF_INITIAL'/`NETWORK_REQUEST_BACKOFF_MAX`) if this causes problems of retrying not fast enough in the beginning or too fast.

Reply:
Thanks for your suggestion. I think the initial and max back-off time could be a follow-up ticket.

Reply:
> Alright, I think this should work in most cases. As a follow-up, we might want to make it configurable or add back-off (similar to `NETWORK_REQUEST_BACKOFF_INITIAL`/`NETWORK_REQUEST_BACKOFF_MAX`) if this causes problems of retrying not fast enough in the beginning or too fast.

Thanks, @azagrebin ! Sorry to jump in, and this is a really good improvement, we have suffered such pains that the K8s API Server hangs after attacked by the infinite creation request from `KubernetesResourceManager`, the whole K8s cluster is unstable in such cases.

Reply:
Thanks for the feedback @zhengcanbin 
We might want to investigate your experience. Previously, we did not have the retry discussed in this PR. Therefore, there may be other sources of these infinite creation requests, e.g. from the #onError callback, depending how the k8s client works internally.
If there is already an issue for your problem (or you could create), could you link to https://issues.apache.org/jira/browse/FLINK-17127 or share your problem there? We might want to have a back-off also for the #onError callback depending on the actual problem.

Reply:
@azagrebin Thanks for the reply! I was planning to dive into this problem to find the root cause and come up with a solution for it. I will try to reproduce this kind of problem, then file an issue, if necessary, to track it and link it to [FLINK-17127](https://issues.apache.org/jira/browse/FLINK-17127), meanwhile, give you more feedback.

Reply:
Sounds good, thank you @zhengcanbin !

Reply:
Hi @azagrebin! I reproduce this problem and create two tickets to track some of the problems I recently found in the `KubernetesResourceManager`. Looking forward to your feedback.
[FLINK-17176](https://issues.apache.org/jira/browse/FLINK-17176)
[FLINK-17177](https://issues.apache.org/jira/browse/FLINK-17177)